Refactor end2 end

snargleplax opened the pull request:
For https://github.com/go-lintpack/lintpack/issues/42.

The new closure within `checkFiles` isn't the most elegant thing ever; let me know if you'd prefer that I try and smooth that out some more before proceeding.

snargleplax committed:
Simplify range loop.

As long as we're ranging over a map, use the form that gives us the
values as well rather than looking them up by indexing into the map.

snargleplax committed:
Invert conditional to nix continue.

Since this is just a one-liner, let's favor brevity.

snargleplax committed:
Rename variable.

Not sure what `y` was meant to stand for, but `w` seems clearer. It's
a warning.

snargleplax committed:
Decouple newWarnings from file IO.

Refactor to just take an io.Reader. Wrap calling code in closure to
provide tighter scope for defer.

snargleplax committed:
Simplify slice handling.

We can just do straightforward assignments here; no need for fancier
slice moves.

snargleplax committed:
Move match tracking out of warning struct.

Having this mutable bit in here is forcing us to do this whole
slice-of-pointers thing that feels like a code smell. Let's take this
path and see if we can simplify things overall.

snargleplax committed:
Nix warning type; just use strings.

Since there's nothing but the text here now, we don't need this
struct.

snargleplax committed:
Pull matched map out of warnings struct.

This seems better with less indirection. Doesn't need to be part of
the struct. This is state for the algorithm, not for the model
representation; let's keep those concerns separate.

snargleplax committed:
Move checkUnmatched.

Since this is now just a helper function that's used in one place,
let's move it closer to that spot.

snargleplax committed:
Change warnings to map type.

Nix struct wrapper, reducing indirection.

cristaloleg reviewed (commented):
SGTM

quasilyte commented:
Is test failure related? Can you reproduce it locally?

snargleplax commented:
Yes and yes; broken by a54e855. Sorry for not catching it. I'll sort it out and work up a fix.

snargleplax committed:
Fix pointer alias bug in warnings.find

The returned pointer needs to point to the actual data within the
slice, since we rely on that identity later in checkUnmatched.

snargleplax commented:
PTAL

quasilyte reviewed (approved):
:+1:

quasilyte commented on linttest/linttest.go:
```diff
@@ -95,27 +96,41 @@ func checkFiles(t *testing.T, c *lintpack.Checker, ctx *lintpack.Context, pkg *p
+			ws, err = newWarnings(rc)
+			if err != nil {
+				t.Fatal(err)
+			}
```
I've used `readfile` before to avoid messing with file closing.
If you do the same here and create a reader out of `[]byte`, closure can be removed and no defers are needed.

snargleplax replied:
Yes, but I was hoping to avoid copying all the bytes an extra time just for the sake of that.

If I just factor out the loop instead (change `checkFiles` to `checkFile`), the closure's not needed and the whole thing gets less indented. I'll push a commit with that and you can see what you think.

snargleplax committed:
Factor out loop and rename to checkFile.

By removing the loop, we can also do away with the closure that
existed to restrict scope for defer.

snargleplax committed:
Use bufio in newWarnings.

Use buffered IO so we don't have to read out all the bytes from the
Reader at once. Also just seems preferable to operate at this level of
abstraction.

snargleplax commented:
PTAL. Speaking of avoiding extra byte copies reminded me to use `bufio` in `newWarnings` too, so I added that as well. Don't know that there's a memory bottleneck there, but it shouldn't hurt. :)

quasilyte commented:
Looks good to me.
@cristaloleg, what do you think?

cristaloleg commented:
Merge?

quasilyte commented:
:tada:
